libs/server/repository/util/src/lib/timeline/generate-stats.util.ts

```diff
@@ -0,0 +1,37 @@
+import { TimelineBuckets } from '@pimp-my-pr/server/repository/core/domain';
+import { TimelineDivisionBase } from '@pimp-my-pr/server/repository/core/domain';
+import { TimelineRecord } from '@pimp-my-pr/server/repository/core/domain';
+import { ITimeTrackable } from '@pimp-my-pr/server/repository/core/domain';
+
+export function generateStats<T extends ITimeTrackable>(
+  buckets: TimelineBuckets<T>,
+  bucketRanges: TimelineDivisionBase,
+  raw: T[]
+): TimelineRecord[] {
+  return Object.entries(buckets).map(
+    ([ts, entry]): TimelineRecord => {
+      const periodEnd = bucketRanges[ts].dateTo.getTime();
+      const periodStart = bucketRanges[ts].dateFrom.getTime();
+      const entityCount = entry.length;
+      const closedBefore = raw.filter(el => el.closedAt && el.closedAt.getTime() < periodStart)
+        .length;
+      const openedAfter = raw.filter(el => el.createdAt.getTime() > periodEnd).length;
+      const entitySumTimeIn = entry.reduce((total, item) => total + item.timeIn, 0);
```

Comment:
Maybe some of this array iterations can be merged into single forEach iteration.